streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -253,17 +293,146 @@ public void buildAndOptimizeTopology() {
         }
     }
 
+    private void maybePerformOptimizations(final Properties props) {
+
+        if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+            maybeOptimizeRepartitionOperations();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void maybeOptimizeRepartitionOperations() {
+
+        for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> entry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            final StreamsGraphNode keyChangingNode = entry.getKey();
+
+            if (entry.getValue().isEmpty()) {
+                continue;
+            }
+
+            final SerializedInternal serialized = new SerializedInternal(getRepartitionSerdes(entry.getValue()));
+
+            final StreamsGraphNode optimizedSingleRepartition = createRepartitionNode(keyChangingNode.nodeName(),
+                                                                                      serialized.keySerde(),
+                                                                                      serialized.valueSerde());
+
+            // re-use parent buildPriority to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setBuildPriority(keyChangingNode.buildPriority());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : entry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
```

Comment:
Actually, Matthias's question is still nagging me... Do we know it's impossible to have two paths back to the keyChangingNode?

Reply:
I think so, but I'll write a test to verify.

Reply:
>Actually, Matthias's question is still nagging me... Do we know it's impossible to have two paths back to the keyChangingNode

Well if the key changing node has multiple children then yes, but code currently accounts for this and unless I misunderstand your concern, that's the main focus of this optimization,  taking multiple child nodes which require a repartition and reducing them down to one where possible.

Having said that, from thinking over the weekend I did come up with an edge condition concerning using `merge` which I'll address as general comment on this PR.

Reply:
Ok, I think that's what I was talking about. Not multiple child nodes, but multiple parents somewhere on the chain back from `repartitionNodeToBeReplaced` to `keyChangingNode`. 

This line of code seems to assume that there is always one path back to `keyChangingNode`, but in general, you could have a diamond (since this is a DAG and not a tree). like :

```
keyChangingNode
  |        |
  c1       c2
   \       /
  (something)
       |
repartitionNodeToBeReplaced
```

If this can occur, then there would need to be multiple of `keyChangingNodeChild` which all need to be re-rooted.

But I'm not sure it can occur.

Reply:
What you are talking about above is the "edge condition" (maybe not an edge condition, but rare?), I was referring to above.

This is the case I was thinking of :

```
updatedA = originalStream.map(...);
updatedB = otherStream.map(...);
merged = updatedA.merge(updatedB)
merged.groupByKey().windowedBy(....)
merged.groupByKey().windowedBy(...)
```
As the code stands right now, we only grab the first key-changing parent node, but in reality, we need to grab both key-changing parent nodes and insert a new repartition node as a child of both.

I'm planning on updating the PR to handle this case.

Reply:
Now seeing the example (super helpful) I am wondering if we can/should be more fancy. I think, it's actually three cases:

```
rekeyed = stream1.map();
merged = rekeyed.merged(stream2);
merged.groupByKey()...
```
For this case, without optimization, we would insert repartition _after_ merge _before_ groupBy -- this mean, we repartition stream2 for no good reason. Could we repartition _before_ merge? Checking first parent might catch this case?

```
rekeyed = stream1.map();
merged = stream2.merged(rekeyed); // similar to above put change order of childen
merged.groupByKey()...
```
This case is similar, but we might not detect if, if we don't check all parent nodes. Thus, we might miss to insert an repartition topic at all in current code?

```
rekeyed1 = stream1.map();
rekeyed2 = stream2.map();
merged = rekeyed1.merged(rekeyed2);
merged.groupByKey()...
```
For this case, we should do the repartition _after_ merge -- otherwise, we would create two repartition topics what is not desired (for the case, that `rekeyed1` and `rekeyed2` are not used elsewhere, too). If `rekeyed1` or `rekeyed2` are use like this:
```
rekeyed1 = stream1.map();
rekeyed1.groupByKey()
rekeyed2 = stream2.map();
merged = rekeyed1.merged(rekeyed2);
merged.groupByKey()...
```
we would still need two repartition topics, because we need content of `stream1` not be mixed with `stream2` in line 2. Also, as `rekeyed1` will be repartitions already before merge(), we don't need to repartition again but `stream2` should be repartition on it's own before merge(), too.

Does this make sense? (note, that merge requires that key and value type of both input streams are the same, and thus, we can use same Serdes independent where we insert the repartition topic(s))